src/software/geom/algorithms/end_in_obstacle_sample.cpp

```diff
@@ -0,0 +1,88 @@
+#include "software/geom/algorithms/end_in_obstacle_sample.h"
+
+#include <random>
+
+#include "software/ai/navigator/obstacle/obstacle.hpp"
+#include "software/geom/point.h"
+
+
+std::optional<Point> endInObstacleSample(const std::vector<ObstaclePtr> obstacles,
+                                         const Point &point,
+                                         const Rectangle &navigable_area,
+                                         int initial_count, double rad_step,
+                                         int per_rad_step, double range)
+{
+    bool point_in_obstacle = false;
+    for (auto const &obstacle : obstacles)
+    {
+        if (obstacle->contains(point) || !contains(navigable_area, point))
+        {
+            if (!point_in_obstacle)
+            {
+                point_in_obstacle = true;
+            }
+
+            Point closest_point            = obstacle->closestPoint(point);
+            bool closest_point_in_obstacle = false;
+            for (auto const &obstacle : obstacles)
+            {
+                if (obstacle->contains(closest_point))
+                {
+                    closest_point_in_obstacle = true;
+                    break;
+                }
+                if (!contains(navigable_area, closest_point))
+                {
+                    closest_point_in_obstacle = true;
+                    break;
+                }
+            }
+
+            if (closest_point_in_obstacle)
+            {
+                break;
+            }
+            else
+            {
+                return closest_point;
+            }
+        }
+    }
+
+    if (!point_in_obstacle)
+    {
+        return point;
+    }
+
+    // sample if original point or closest point outside initial obstacle checked aren't
+    // valid
+    double rad        = 0.15;
+    int steps_per_rad = 6;
+    while (rad <= range)
+    {
+        double increment = 360.0 / steps_per_rad;
+        for (int i = 0; i < steps_per_rad; i++)
+        {
+            Angle angle        = Angle::fromDegrees(static_cast<float>(i) * increment);
+            Vector direction   = Vector::createFromAngle(angle);
+            Point sample_point = point + direction * rad;
+            bool sample_point_in_obstacle = false;
+            for (auto const &obstacle : obstacles)
+            {
+                if (obstacle->contains(sample_point) ||
+                    !contains(navigable_area, sample_point))
+                {
+                    sample_point_in_obstacle = true;
+                    break;
+                }
+            }
+            if (!sample_point_in_obstacle)
+            {
+                return sample_point;
+            }
+        }
```

Comment:
Unless I'm confused, the objective of this function is to compute the nearest boundary point from an obstacles centroid given a set of obstacles, if the given point is contained within any of those obstacles. There's a potentially a different way to approach that:

Start by sorting the list of obstacles at the start of the call, based on their squared distance to the point, as well as any boundary points if they exist given the condition. If none of the obstacles contain it, return null, but if they do you can merge the offending obstacles that do contain it recursively with the following algorithm (that in my mind sound like it might work?):

The implementation would be similar to `intersection.h` for how the method `intersection` is overloaded to compute the intersection of a polygon and a segment, so you would just need to extend that to loop through another polygon's segments. Lastly, you would need to combine the segment lists and replace the intersecting segments with new segments to and from the intersection points.